File: python/53_Maximum_Subarray.py
#!/usr/bin/python
import sys
import os

'''
For example, given the array [-2,1,-3,4,-1,2,1,-5,4],
the contiguous subarray [4,-1,2,1] has the largest sum = 6.
'''

class Solution(object):
    def maxSubArray(self, nums):
		# dp[i] : the sum of the maximum subarray ended at i
		n = len(nums)
		if not n:
			return None
		dp = [0 for i in range(n)]
		dp[0] = nums[0]
		maxSum = nums[0]
		for i in range(1, n):
			dp[i] = dp[i - 1] + nums[i] if dp[i - 1] > 0 else nums[i]
			maxSum = max(maxSum, dp[i])
		return maxSum
		


def main():
    s = Solution()
    nums = [-2,1,-3,4,-1,2,1,-5,4]
    print s.maxSubArray(nums)

if __name__ == "__main__":
    main()

File: c++/53_Maximum_Subarray.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public: // dp[i] : sum of maximum value ended at i
    int maxSubArray(vector<int>& num) {
    	int n = num.size();
    	if (n == 0)
    		return NULL;
    	int maxS = num[0];
    	int dp[n] = {0};
    	dp[0] = num[0];
    	for (int i = 1; i < n; i++){
    		dp[i] = dp[i - 1] > 0 ? dp[i - 1] + num[i] : num[i];
    		maxS = max(maxS, dp[i]);
    	}
    	return maxS;
    }
};

